x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java

```diff
@@ -339,4 +342,30 @@ void verifyInternalConsistency() {
             }
         });
     }
+
+    static BitSet bitSetFromDocIterator(DocIdSetIterator iter, int maxDoc) throws IOException {
+        // TODO: This snippet is copied from Lucene Bitset#of. Should we integrate it to Lucene?
+        final long cost = iter.cost();
+        final int threshold = maxDoc >>> 7;
+        final BitSet set;
+        if (cost < threshold) {
+            set = new SparseFixedBitSet(maxDoc);
+        } else {
+            set = new FixedBitSet(maxDoc);
+        }
+        if (iter.docID() != -1) {
+            throw new IllegalStateException("Must be an unpositioned iterator, got current position = " + iter.docID());
+        }
+        int matches = 0;
+        for (int doc = iter.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = iter.nextDoc()) {
+            matches++;
+            set.set(doc);
+        }
+        if (matches == maxDoc) {
```

Comment:
The BitSet#or implementations seem to have some optimizations, maybe we should do set.or(it) and then compare set.cardinality() with maxDoc.